Use wayland-scanner at build time instead of pre-generated files

vadz opened the pull request:
Should fix problems due to the pre-generated files in the repository being incompatible with the build system Wayland version. It also allows to drop some hacks that were put in place to deal with such incompatibilities before.

@MaartenBent I'm not sure if using an `ALL` custom target is really a good idea, maybe we should just generate these files during configure step, what do you think?

vadz committed:
Add wxUSE_WAYLAND option allowing to disable Wayland support

Allow disabling Wayland support by passing --without-wayland option to
configure or -DwxUSE_WAYLAND=0 to CMake if it is undesirable for
whatever reason.

vadz committed:
Generate Wayland protocol code using wayland-scanner at build time

Make sure that the generated code is compatible with the system we're
building on, which could not be the case before with pre-generated
files (that are now removed from the repository).

See #25721.

Closes #25768.

vadz committed:
Revert "Exclude headers generated by wayland-scanner from code checks"

This reverts commit c375a91dd6b3129f40ce701485d472d8eda43daa as it's
not needed any longer because the files in question are not present in
the repository any more.

vadz committed:
Revert "Also exclude wayland-scanner-generated sources from space checks"

This reverts commit 63707e27679d8da70ba865da2c4561905daa4904 for the
same reason as in the previous commit: the generated sources are not
stored in the repository any longer.

vadz committed:
Revert "Fix compilation with old Wayland client protocol headers"

This reverts commit ceace944b275631aae631766b4c9bf6bb37caaa4 because the
pointer-wrap protocol header generated on the build system should be
compatible with its Wayland version and so this hack is not needed any
more.

vadz committed:
Explicitly create protocols output directly for CMake/make

Ninja is smart enough to create output directories on its own, but when
using make generator we need to do it ourselves.

vadz committed:
Fix wrong test for WAYLAND_CLIENT in CMake

Correct the test added in 147458e0ed (Use wp_pointer_warp_v1 if
available in wxWindow::WarpPointer(), 2025-07-28) which was missing the
module to check for.

vadz committed:
Improve handling of wayland-client in CMake

Add only WAYLAND_CLIENT_INCLUDE_DIRS to wxTOOLKIT_INCLUDE_DIRS instead
of adding WAYLAND_CLIENT_CFLAGS to them as CFLAGS might, in theory,
contain something else.

For now just ignore any non -I options from CFLAGS (there shouldn't be
any, normally).

vadz committed:
Warn if wayland-client or wayland-scanner are not found

If these warnings are annoying, Wayland support can be explicitly
disabled to avoid them.

MaartenBent commented:
Indeed, probably better to create them during configuring. I think that is what `configure / AC_CONFIG_COMMANDS` does too? 

If not, you'll need to add an dependency of `core` to this custom target, because these files need to be created before building `core`. In the `core` cmake file something like:
```cmake
wx_add_dependencies(wxcore wayland_protocols)
```

I haven't checked the details of this (or the previous) PR, but does the generated header need to be installed as well or is it only used privately when building?

Most of the `wxUSE_*` variables also have an entry in `setup.h[.in]`. Would that be useful here too?

vadz commented:
> Indeed, probably better to create them during configuring. I think that is what `configure / AC_CONFIG_COMMANDS` does too?

Yes, it does, but it also allows you to recreate them later by running `./config.status wayland-protocols` in the build directory, which is why I wanted to have a matching CMake target. But as Wayland XML files change very rarely, if ever, it's probably not necessary after all.
 
> If not, you'll need to add an dependency of `core` to this custom target

This is not really relevant any more, but I thought that `ALL` custom targets were made before anything else, weren't they?

> I haven't checked the details of this (or the previous) PR, but does the generated header need to be installed as well or is it only used privately when building?

No, it's only used privately.
 
> Most of the `wxUSE_*` variables also have an entry in `setup.h[.in]`. Would that be useful here too?

I decided it wasn't, as I don't see what would it be useful for, but if anybody thinks we need it we can always do it later.

I suspect that [this CI failure](https://github.com/wxWidgets/wxWidgets/actions/runs/17534666873/job/49796172463?pr=25769) is due to `WAYLAND_CLIENT_CFLAGS` containing something other than definitions, so I'm probably going to have to revert this commit...

vadz committed:
Remove wayland-protocols CMake custom command

Generate the required files during the configure step instead, we don't
really need to ever regenerate them, so this is simpler and doesn't take
any time when building later.

Blake-Madden commented:
@vadz Success! Using this PR's branch in my GH action fixed it. Thanks!
